tools/osv-linter/internal/checks/checks.go

```diff
@@ -0,0 +1,140 @@
+// Package checks defines and implements all checks and collections of checks.
+//
+// To add additional checks:
+// 1. define a new instance of `Check`
+// 2. add it to the `checks` array
+// 3. add it to the relevent collections defined in `checkCollections`
+//
+// To add additional collections of checks:
+// 1. add to the `checkCollections` array.
+package checks
+
+import (
+	"fmt"
+
+	"github.com/tidwall/gjson"
+)
+
+// A CheckCode is a unique code for a check.
+type CheckCode string
+
+// CheckError describes when a check fails.
+type CheckError struct {
+	Code    CheckCode
+	Message string
+}
+
+// Error returns the error message, including the code.
+func (ce *CheckError) Error() string {
+	return fmt.Sprintf("%s: %s", ce.Code, ce.Message)
+}
+
+// CodeString returns just the error code, as a string.
+func (ce *CheckError) CodeString() string {
+	return string(ce.Code)
+}
+
+// Checkers are for running a discrete checking function.
+type Checker interface {
+	CodeString() string
+	Name() string
+	Description() string
+	Run(*gjson.Result) []CheckError
+}
+
+// Check defines a single check.
+type Check struct {
+	code        CheckCode
+	name        string
+	description string
+	check       func(*gjson.Result) []CheckError
+}
+
+// Run runs the check, returning any findings.
+// The check has no awareness of the CheckCode,
+// this merges that with the check's findings.
+func (c *Check) Run(json *gjson.Result) (findings []CheckError) {
+	for _, finding := range c.check(json) {
+		findings = append(findings, CheckError{
+			Code:    c.code,
+			Message: finding.Error(),
+		})
+	}
+	return findings
+}
+
+// Name returns the name of the check.
+func (c *Check) Name() string {
+	return c.name
+}
+
+// Description returns the description of the check.
+func (c *Check) Description() string {
+	return c.description
+}
+
+// CodeString returns the short code for the check, as a string.
+func (c *Check) CodeString() string {
+	return string(c.code)
+}
+
+// Collection is a named collection of checks.
+type Collection struct {
+	name        string
+	description string
+	checks      []*Check
+}
+
+// Name returns the name of the collection.
+func (cc *Collection) Name() string {
+	return cc.name
+}
+
+// Description returns the description of the collection.
+func (cc *Collection) Description() string {
+	return cc.description
+}
+
+// Checks returns the checks in the collection.
+func (cc *Collection) Checks() []*Check {
+	return cc.checks
+}
+
+var CheckIntroducedEventExists = &Check{
+	code:        "R0001",
```

Comment:
nit: should `code` be more human readable? 

the fact that there seems to be some sequencing implied here makes it hard to keep checks more modular -- because they have to have knowledge about other checks that exist to know what number in the line they are.

Reply:
e.g. "introduced-event-exists" as the code.

Reply:
I was modeling this off `pylint`, which has both fixed-length codes and more human-friendly variable-length strings, e.g. from `pylint --list-msgs`:

```
:line-too-long (C0301): *Line too long (%s/%s)*
  Used when a line is longer than a given number of characters.
```